### PR TITLE
Added support for Generic Resources

### PR DESCRIPTION
**Pull request by RenaudWasTaken**
Signed-off-by: Renaud Gaubert <rgaubert@nvidia.com>
Hi !

This implements support for Generic Resources in moby (after implementing it in swarmkit).
More details can be found on the issue [here](https://github.com/moby/moby/issues/33439) and the [design doc](https://github.com/docker/swarmkit/blob/master/design/generic_resources.md)

Thanks,

### CODE REVIEW

File: daemon/cluster/convert/service.go [hunk 1]
```diff
@@ -293,15 +318,33 @@ func resourcesFromGRPC(res *swarmapi.ResourceRequirements) *types.ResourceRequir
...
+				MemoryBytes:      res.Reservations.MemoryBytes,
+				GenericResources: GenericResourcesFromSwarmkit(res.Reservations.Generic),
 			}
 		}
 	}
 
 	return resources
 }
 
+func GenericResourcesToSwarmkit(genericRes []*types.GenericResource) []*swarmapi.GenericResource {
```
Comment (aaronlehmann) [inline daemon/cluster/convert/service.go]: nit: `GenericResourcesToGRPC`

File: daemon/cluster/convert/service.go [hunk 2]
```diff
@@ -281,6 +282,30 @@ func annotationsFromGRPC(ann swarmapi.Annotations) types.Annotations {
 	return a
 }
 
+func GenericResourcesFromSwarmkit(genericRes []*swarmapi.GenericResource) []*types.GenericResource {
```
Comment (aaronlehmann) [inline daemon/cluster/convert/service.go]: nit: `GenericResourcesFromGRPC`

File: api/types/swarm/task.go [hunk 3]
```diff
@@ -71,8 +72,24 @@ type TaskSpec struct {
...
-	NanoCPUs    int64 `json:",omitempty"`
-	MemoryBytes int64 `json:",omitempty"`
+	NanoCPUs         int64              `json:",omitempty"`
+	MemoryBytes      int64              `json:",omitempty"`
+	GenericResources []*GenericResource `json:",omitempty"`
+}
+
+type GenericResource struct {
+	Str      *GenericString
+	Discrete *GenericDiscrete
```
Comment (aaronlehmann) [inline api/types/swarm/task.go]: Both of these fields should have `json:",omitempty"`

**Review by aaronlehmann — COMMENTED** [review]
The design seems right to me. I agree with configuring the available resources on the node itself instead of on a manager, because that leaves open the future possibility of automatically detecting devices such as GPUs. It also fits the existing model where the node self-reports memory and CPU resources.

In the future we could potentially also add generic resource information to `NodeSpec`, so certain resources could be declared with `node update` on a manager. This could coexist with individual nodes self-reporting any resources they know about. The `NodeSpec` model might make sense for "abstract" resources (as opposed to real hardware resources). However, I think the self-reporting approach in this PR is the right place to start.

I think the main questions around this will center around UX. I'm not an expert in that area so I'll let others review this part and provide suggestions.

**Comment by RenaudWasTaken** [comment]
@aaronlehmann I've addressed your issues, fixed vendors (I think) and compilation.
Should I also update api/swagger.yaml ?

**Comment by aaronlehmann** [comment]
> @aaronlehmann I've addressed your issues, fixed vendors (I think) and compilation.
> Should I also update api/swagger.yaml ?
> 
Yes, please.

**Comment by RenaudWasTaken** [comment]
@aaronlehmann I've create a `ResourceObject` object to encapsulate `NanoCPUs` and `MemoryBytes` which were hardcoded in every structures.

I'm not really sure about the choice of name but because there already exists a `Resources` object I don't have any other suggestions. Do you have some idea ?

Thanks!

**Comment by RenaudWasTaken** [comment]
@aaronlehmann @stevvooe @ehazlett @tonistiigi PTAL :)
Thanks!

**Comment by tiborvass** [comment]
I wonder how this impacts https://github.com/moby/moby/pull/33575 by @cpuguy83

**Comment by RenaudWasTaken** [comment]
@tiborvass I gave the PR a quick look and it seems to me that it wouldn't impact it :)
Though I think that the current model would let your plugin request Generic resources which I think is something that seems reasonable ?

**Comment by cpuguy83** [comment]
Agreed I don't think there is any impact.

File: vendor.conf [hunk 4]
```diff
@@ -104,7 +104,7 @@ github.com/containerd/containerd 3addd840653146c90a254301d6c3a663c7fd6429
 github.com/tonistiigi/fifo 1405643975692217d6720f8b54aeee1bf2cd5cf4
 
 # cluster
-github.com/docker/swarmkit eb07af52aa2216100cff1ad0b13df48daa8914bf
+github.com/docker/swarmkit 85019dae9bacf2b68b81f092c34e755e2ee76b02
```
Comment (tophj-ibm) [inline vendor.conf]: Hey @RenaudWasTaken 👋 

You should be able to rebase and drop the swarmkit re-vendor now that a newer version is in moby via https://github.com/moby/moby/pull/33642.

Comment (RenaudWasTaken) [inline vendor.conf]: Thanks !

**Comment by RenaudWasTaken** [comment]
This PR seems to be in limbo :)
Is there any ETA or process to follow @aaronlehmann @stevvooe @ehazlett @tonistiigi @aluzzardi  ?

Thanks!

File: api/types/swarm/task.go [hunk 5]
```diff
@@ -71,8 +72,30 @@ type TaskSpec struct {
...
-	MemoryBytes int64 `json:",omitempty"`
+	NanoCPUs         int64              `json:",omitempty"`
+	MemoryBytes      int64              `json:",omitempty"`
+	GenericResources []*GenericResource `json:",omitempty"`
+}
+
+// GenericResource represents a "user defined" resource which can
+// be either an integer (e.g: SSD=3) or a string (e.g: SSD=sda1)
+type GenericResource struct {
+	Str      *GenericString   `json:",omitempty"`
```
Comment (cpuguy83) [inline api/types/swarm/task.go]: What does `Str` mean? I assume "string" as in "GenericString", but it's not very descriptive.

File: api/types/swarm/task.go [hunk 6]
```diff
@@ -71,8 +72,30 @@ type TaskSpec struct {
...
+	Discrete *GenericDiscrete `json:",omitempty"`
+}
+
+// GenericString represents a "user defined" resource which is defined
+// as a string
+type GenericString struct {
+	Kind  string `json:",omitempty"`
```
Comment (cpuguy83) [inline api/types/swarm/task.go]: What consumes these "kind" fields? Some extra docs would be helpful here.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: api/types/swarm/task.go (hunk 6 above)
Comment (RenaudWasTaken) [inline api/types/swarm/task.go]: Hi !

Kind is used to describe the Kind of the resource (e.g: GPU, FPGA, ...)
It is then used when creating a service ($ docker service create --name nginx --image nginx:latest --generic-resources "GPU=2")


You can find the design doc here: https://github.com/docker/swarmkit/blob/master/design/generic_resources.md

File: api/types/swarm/task.go (hunk 5 above)
Comment (RenaudWasTaken) [inline api/types/swarm/task.go]: Yep Str means string.
A resource can either be described by an int (SSD=3) or through a String (GPU=UUID-1)

This is described in more details in the design doc: https://github.com/docker/swarmkit/blob/master/design/generic_resources.md#use-cases

Comment (RenaudWasTaken) [inline api/types/swarm/task.go]: Do you have a more descriptive name in mind ?

File: api/types/swarm/task.go (hunk 6 above)
Comment (cpuguy83) [inline api/types/swarm/task.go]: So this is completely opaque and dependent on how the resource is defined by the engine administrator.
Can we add some doc around that?

File: api/types/swarm/task.go (hunk 5 above)
Comment (cpuguy83) [inline api/types/swarm/task.go]: That's a great question...


Maybe `NamedResourceSpec`
And likewise `DiscreteResourceSpec`.

Although I must admit I don't know what the idea is behind the naming of `Discrete`

File: api/types/swarm/task.go [hunk 7]
```diff
@@ -71,8 +72,30 @@ type TaskSpec struct {
...
+// GenericResource represents a "user defined" resource which can
+// be either an integer (e.g: SSD=3) or a string (e.g: SSD=sda1)
+type GenericResource struct {
+	Str      *GenericString   `json:",omitempty"`
+	Discrete *GenericDiscrete `json:",omitempty"`
+}
+
+// GenericString represents a "user defined" resource which is defined
+// as a string
+type GenericString struct {
```
Comment (cpuguy83) [inline api/types/swarm/task.go]: In following with the above suggesting, `NamedGenericResource`

File: api/types/swarm/task.go [hunk 8]
```diff
@@ -71,8 +72,30 @@ type TaskSpec struct {
...
+	Kind  string `json:",omitempty"`
+	Value string `json:",omitempty"`
+}
+
+// GenericDiscrete represents a "user defined" resource which is defined
+// as an integer
+type GenericDiscrete struct {
```
Comment (cpuguy83) [inline api/types/swarm/task.go]: `DiscreteGenericResource`

File: daemon/cluster/executor/container/container.go [hunk 9]
```diff
@@ -186,13 +187,16 @@ func (c *containerConfig) exposedPorts() map[nat.Port]struct{} {
...
+	env := append(c.spec().Env, genericEnvs...)
+
 	config := &enginecontainer.Config{
 		Labels:       c.labels(),
 		StopSignal:   c.spec().StopSignal,
 		Tty:          c.spec().TTY,
 		OpenStdin:    c.spec().OpenStdin,
 		User:         c.spec().User,
-		Env:          c.spec().Env,
+		Env:          env,
```
Comment (cpuguy83) [inline daemon/cluster/executor/container/container.go]: How do the devices make it into the container?

Comment (RenaudWasTaken) [inline daemon/cluster/executor/container/container.go]: Isolation, or resource allocation (for example gpu devices) is a totally different scope :)
This PR is about resource accounting at the scheduler level and giving users the knobs.

The problem of GPUs in containers is not an easy one and right now we are solving it with a library that can be called at the runc level: https://github.com/NVIDIA/libnvidia-container

Comment (cpuguy83) [inline daemon/cluster/executor/container/container.go]: Ah, ok. Does it make sense to expose this to the container then?

Comment (RenaudWasTaken) [inline daemon/cluster/executor/container/container.go]: Well even if it's not the scope of this PR to isolate or allocate the resource, you still want at some point the container to know what resources it was assigned.

Right now we are doing this through environment variables and this allows to send the information to the runtime level.

**Comment by RenaudWasTaken** [comment]
@cpuguy83 I've addressed your comments :)

**Comment by RenaudWasTaken** [comment]
Rebased against master seems like this broke the WindowsRS1 jenkins :(
@cpuguy83 do you have any more comments or does this seem good to you ?

@stevvooe @ehazlett @tonistiigi can you take a look :) ?

File: api/swagger.yaml [hunk 10]
```diff
@@ -487,6 +487,41 @@ definitions:
...
+      GenericResources:
+        $ref: "#/definitions/GenericResources"
+
+  GenericResources:
+    description: "User defined Resources, can be either Integer resources (e.g: SSD=3) or String resources (e.g: GPU={UUID1, UUID2})"
+    type: "array"
+    items:
+      type: "object"
+      properties:
+        NamedResourceSpec:
```
Comment (stevvooe) [inline api/swagger.yaml]: This looks very different from what is in swarmkit. https://github.com/docker/swarmkit/blob/d4a01ccd09df3e554d9efc670869fd130ed0154e/api/types.proto#L38

File: api/types/swarm/task.go [hunk 11]
```diff
@@ -40,13 +40,14 @@ type Task struct {
...
-	DesiredState        TaskState           `json:",omitempty"`
-	NetworksAttachments []NetworkAttachment `json:",omitempty"`
+	Spec                     TaskSpec            `json:",omitempty"`
+	ServiceID                string              `json:",omitempty"`
+	Slot                     int                 `json:",omitempty"`
+	NodeID                   string              `json:",omitempty"`
+	Status                   TaskStatus          `json:",omitempty"`
+	DesiredState             TaskState           `json:",omitempty"`
+	NetworksAttachments      []NetworkAttachment `json:",omitempty"`
+	AssignedGenericResources []*GenericResource  `json:",omitempty"`
```
Comment (stevvooe) [inline api/types/swarm/task.go]: This field name is extremely awkward. Can there every be unassigned resources? These resources belong to the task and are assumed to be assigned.

Comment (stevvooe) [inline api/types/swarm/task.go]: Also, why a pointer type?

File: api/types/swarm/task.go [hunk 12]
```diff
@@ -71,8 +72,34 @@ type TaskSpec struct {
 
 // Resources represents resources (CPU/Memory).
 type Resources struct {
-	NanoCPUs    int64 `json:",omitempty"`
-	MemoryBytes int64 `json:",omitempty"`
+	NanoCPUs         int64              `json:",omitempty"`
+	MemoryBytes      int64              `json:",omitempty"`
+	GenericResources []*GenericResource `json:",omitempty"`
```
Comment (stevvooe) [inline api/types/swarm/task.go]: No pointer type here.

File: api/swagger.yaml (hunk 10 above)
Comment (cpuguy83) [inline api/swagger.yaml]: Sorry, this was my suggestion just reading through the API changes in this PR.

**Comment by stevvooe** [comment]
So, let's make sure the names match in swarmkit. Either backport these object names to swarmkit or change these to match swarmkit. It makes it extremely hard to understand what is equivalent when they have different names.

File: api/types/swarm/task.go (hunk 11 above)
Comment (RenaudWasTaken) [inline api/types/swarm/task.go]: No you can't have unassigned resources that case do you think `GenericResources` is enough ?

It should not be a pointer type will change that

File: api/types/swarm/task.go (hunk 12 above)
Comment (RenaudWasTaken) [inline api/types/swarm/task.go]: will change that

**Comment by RenaudWasTaken** [comment]
I'll do a PR to swarmkit I think the naming @cpuguy83 proposed was better

File: api/types/swarm/task.go (hunk 11 above)
Comment (stevvooe) [inline api/types/swarm/task.go]: `Resources` might even be okay. We'd have to flow some of this back to swarmkit (these names are a little clearer).

@aaronlehmann Thoughts? I know this has already been through a large review cycle.

**Comment by stevvooe** [comment]
@RenaudWasTaken I agree! Thank you!

**Comment by RenaudWasTaken** [comment]
@stevvooe once the PR gets accepted in swarmkit should I do a seperate PR to update moby dependencies or should I update the deps in this PR ?

**Comment by stevvooe** [comment]
@RenaudWasTaken Whatever is easiest for you!

**Comment by RenaudWasTaken** [comment]
@stevvooe Rebased against master with the latests swarmkit revendoring

**Comment by cpuguy83** [comment]
This looks pretty good to me.

The only concern I have is allowing the engine to define the resources (via the info endpoint as it does in in this PR) vs doing it at the node level.
By "concern" here, I mean I'm not sure which is the right approach, not necessarily that I think there's something wrong with it.

**Comment by RenaudWasTaken** [comment]
> The only concern I have is allowing the engine to define the resources (via the info endpoint as it does in in this PR) vs doing it at the node level.

@cpuguy83 what do you mean by doing at the node level ?

**Comment by cpuguy83** [comment]
@RenaudWasTaken Sorry, for "node" level I mean on the swarm node object rather than the engine itself.

**Comment by RenaudWasTaken** [comment]
@cpuguy83 we discussed that with @aaronlehmann before so I'll be quoting him :)

> The current behavior for the existing resources (NanoCPUs and MemoryBytes) is that they get filled in by Describe (see agent/exec/dockerapi/executor.go for the dockerapi executor implementation). I think we should follow this with generic resources, so that they are consistent with the existing resources, and because it's an easy way to address the hotplug case (Describe is called periodically, and the node is updated dynamically based on the result).
 
and:

> But once this is all integrated in Docker, I think the source of truth should be the SystemInfo call, not information passed through Node.

**Comment by RenaudWasTaken** [comment]
@cpuguy83 @stevvooe is there anything in this PR that needs to be changed ?

**Comment by stevvooe** [comment]
LGTM

File: daemon/cluster/convert/service.go [hunk 13]
```diff
@@ -313,15 +339,34 @@ func resourcesFromGRPC(res *swarmapi.ResourceRequirements) *types.ResourceRequir
...
 
+// GenericResourcesToGRPC converts a GenericResource to a GRPC GenericResource
+func GenericResourcesToGRPC(genericRes []types.GenericResource) []*swarmapi.GenericResource {
+	var generic []*swarmapi.GenericResource
+	for _, res := range genericRes {
+		var r *swarmapi.GenericResource
+
+		if res.DiscreteResourceSpec != nil {
+			r = genericresource.NewDiscrete(res.DiscreteResourceSpec.Kind, res.DiscreteResourceSpec.Value)
+		} else {
```
Comment (aaronlehmann) [inline daemon/cluster/convert/service.go]: I think this should be `else if res.NamedResourceSpec != nil`. These will be provided by users, and we can't assume that one of the two is non-nil.

Comment (RenaudWasTaken) [inline daemon/cluster/convert/service.go]: Fixed

**Comment by aaronlehmann** [comment]
LGTM

**Comment by RenaudWasTaken** [comment]
CI is now green :)

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by RenaudWasTaken** [comment]
@aaronlehmann @cpuguy83 what is the process to integrate in the CLI ?
It seems they don't rebase very often should I open an issue ?

**Comment by cpuguy83** [comment]
@RenaudWasTaken Rebasing is done as needed. If you have changes that need to go into the CLI that require a bump of moby/moby, you can revendor in 1 commit and make your changes in a 2nd commit.

**Pull request merged**